docs/architecture/adr-050-sign-mode-textual.md

```diff
@@ -0,0 +1,492 @@
+Memo: <string>                                              // Skipped if no memo set
+*Gas Limit: <uint64>
+*Timeout Height:  <uint64>                                  // Skipped if no timeout_height set
+Tipper: <string>                                            // If there's a tip
+Tip: <string>
+*This transaction has <int> body extension:                 // Skipped if no body extension options
+*<repeated Any>
+*This transaction has <int> body non-critical extensions:   // Skipped if no body non-critical extension options
+*<repeated Any>                                             // See value renderers for Any and array encoding.
+*This transaction has <int> body auth info extensions:      // Skipped if no auth info extension options
+*<repeated Any>
+*This transaction has <int> other signers:                  // Skipped if there is only one signer
+*Signer (<int>/<int>):
+*Public Key: <base64_string>
+*Sequence: <uint64>
+*End of other signers
+```
+
+### 8. Encoding of the Transaction Body
+
+We call "transaction body" the `Tx.TxBody.Messages` field, which is an array of `Any`s. Since messages are widely used, they have a slightly different encoding than usual array of `Any`s (protobuf: `repeated repeated google.protobuf.Any`) described in Annex 1.
+
+```
+This transaction has <int> message:   // Optional 's' for "message" if there's is >1 sdk.Msgs.
+// For each Msg, print the following 2 lines:
+Msg (<int>/<int>): <string>           // E.g. Msg (1/2): bank v1beta1 send coins
+<value rendering of Msg struct>
+End of transaction messages
+```
+
+#### Example
+
+Given the following Protobuf message:
+
+```proto
+message Grant {
+  google.protobuf.Any       authorization = 1 [(cosmos_proto.accepts_interface) = "Authorization"];
+  google.protobuf.Timestamp expiration    = 2 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
+}
+
+message MsgGrant {
+  option (cosmos.msg.v1.signer) = "granter";
+  option (cosmos.msg.v1.textual) = "authz v1beta1 grant";
+
+  string granter = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  string grantee = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+}
+```
+
+and a transaction containing 1 such `sdk.Msg`, we get the following encoding:
+
+```
+This transaction has 1 message:
+Msg (1/1): authz v1beta1 grant
+Granter: cosmos1abc...def
+Grantee: cosmos1ghi...jkl
+End of transaction messages
+```
+
+### 9. Wire Format
+
+This string array is encoded as a single `\n`-delimited string before transmitted to the hardware device.
+
+## Additional Formatting by the Hardware Device
+
+Hardware devices differ in screen sizes and memory capacities. The above specifications are all verified on the protocol level, but we still allow the hardware device to add custom formatting rules that are specific to the device. Rules can include:
+
+- if a string is too long, show it on multiple screens,
+- break line between the `key` and `value` from #3,
+- never allow line breaks on a number or a coin value.
```

Comment:
if there is a big number, then how a user will know the value if we can't scroll and can't break lines?

Reply:
up, this is still a concern.

Reply:
Addressed. The way is to do best effort:
- if small number, no problem
- if big number, then show it on separate line
- if huge number (rare case), show it on multiple screens

Reply:
maybe we can have an exponential notation? For example, instead `12300000000000000` we write `1.23e18`

Reply:
I think there's another thread with scientific notation. My opinion:
- it's IMO not known enough by people with non-scientific background. You never see this notation in bank/finance for example.
- `1234567890` would be written as `1,23456789e9`, which is actually longer

Reply:
I agree. I studied math and CS and consider this `e` notation unreadable in any financial context.

Reply:
I  think it would be good to have some shortcuts. For common Ethereum based coins (with 1e18 base) it will be difficult to read even if we use `_` or `'` every 3 or 6 digits.

Reply:
Don't we have display respresentations to solve this such that amounts are dispplayed in ETH instead of Wei anyways?

Reply:
I think the concern here is that we _always_ display in the display denom, even if it doesn't make sense.

Take 20Wei, and the display denom is ETH. Textual will display `0.00000000000000002 ETH`, even though `20 Wei` is way more readable.

Maybe an idea is to display in the denom with the shortest string length.